### PR TITLE
replaced content_type from DOCString to Mime

### LINKED ISSUE #16483: Document::content_type should be of type Mime

### PR DESCRIPTION
**Pull request by nupurbaghel**
Replaced content_type from DomString to Mime

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #16483 (https://github.com/servo/servo/issues/16483).

- [x] These changes do not require tests because they make changes to already existing files

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19972)

### CODE REVIEW

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -2222,12 +2224,12 @@ impl Document {
             has_browsing_context: has_browsing_context == HasBrowsingContext::Yes,
             implementation: Default::default(),
             content_type: match content_type {
-                Some(string) => string,
-                None => DOMString::from(match is_html_document {
+                Some(string) => string.parse().unwrap(),
```
Comment (jdm) [inline components/script/dom/document.rs]: `unwrap()` here will panic if we ever get a content type that isn't a valid mime type. Let's use `unwrap_or` and pass a default value of text/html.

File: components/script/dom/document.rs [hunk 2]
```diff
@@ -2969,7 +2971,8 @@ impl DocumentMethods for Document {
             local_name.make_ascii_lowercase();
         }
 
-        let ns = if self.is_html_document || self.content_type == "application/xhtml+xml" {
+        let ns = if self.is_html_document || ((self.content_type.0 == TopLevel::Application) &&
+                                               (self.content_type.1.as_str() == "xhtml+xml")) {
```
Comment (jdm) [inline components/script/dom/document.rs]: Let's do this:
```rust
let is_xhtml = self.content_type.0 == TopLevel::Application && self.content_type.1.as_str() == "xhtml+xml";
let ns = if self.is_html_document || is_xhtml {
```

File: components/script/dom/document.rs [hunk 3]
```diff
@@ -2889,7 +2891,7 @@ impl DocumentMethods for Document {
 
     // https://dom.spec.whatwg.org/#dom-document-content_type
     fn ContentType(&self) -> DOMString {
-        self.content_type.clone()
+        DOMString::from_string(self.content_type.0.to_string() + "/" + &self.content_type.1.to_string()).clone()
```
Comment (jdm) [inline components/script/dom/document.rs]: We can do `DOMString::from(self.content_type.as_str())`. We can remove the `.clone()` now, too.

**Review by jdm — CHANGES_REQUESTED** [review]
Good start! Seeing the `unwrap` made me start looking at the various places that call `Document::new`, and I now think that we should change it to accept an `Option<Mime>` value as well (along with `Document::new_inherited`, `XMLDocument::new_inherited`, and `XMLDocument::new`), because almost every single caller I looked at already has a Mime value and converts it to a string just to call `Document::new`.

File: components/script/dom/document.rs (hunk 1 above)
Comment (nupurbaghel) [reply components/script/dom/document.rs]: Perfect, I will do that.

File: components/script/dom/document.rs (hunk 2 above)
Comment (nupurbaghel) [reply components/script/dom/document.rs]: Sure, I get that 👍

**Comment by nupurbaghel** [comment]
> We should change it to accept an Option<Mime> value as well (along with Document::new_inherited, XMLDocument::new_inherited, and XMLDocument::new)

Yeah, sounds perfect. Should we makes these in this PR itself or should I create another PR for it after completing this one?

**Comment by jdm** [comment]
Let's do it as part of this PR.

File: components/script/dom/document.rs (hunk 1 above)
Comment (nupurbaghel) [reply components/script/dom/document.rs]: After updating params of new_inherited, we can directly assign the value of string(like it was previously) without even using unwrap.

Comment (jdm) [reply components/script/dom/document.rs]: Oh, right. Even better!

File: components/script/dom/document.rs (hunk 3 above)
Comment (nupurbaghel) [reply components/script/dom/document.rs]: I recently found that .as_str() method doesn't actually exist for Mime type, is there any other method other than the one I used, which will look better here?

File: components/script/dom/domparser.rs [hunk 4]
```diff
@@ -54,7 +55,7 @@ impl DOMParserMethods for DOMParser {
                        ty: DOMParserBinding::SupportedType)
                        -> Fallible<DomRoot<Document>> {
         let url = self.window.get_url();
-        let content_type = DOMString::from(ty.as_str());
+        let content_type = ty.as_str().parse().unwrap_or(Mime(TopLevel::Text, SubLevel::Html, vec![]));
```
Comment (jdm) [inline components/script/dom/domparser.rs]: All of the supported types should be valid MIME types, so let's do this instead: `ty.as_str().parse().expect("Supported type is not a MIME type")`.

File: components/script/dom/document.rs [hunk 5]
```diff
@@ -2224,7 +2224,7 @@ impl Document {
             has_browsing_context: has_browsing_context == HasBrowsingContext::Yes,
             implementation: Default::default(),
             content_type: match content_type {
-                Some(string) => string.parse().unwrap(),
+                Some(Mime_data) => Mime_data,
```
Comment (jdm) [inline components/script/dom/document.rs]: Our coding style would make this `mime_data`.

**Review by jdm — CHANGES_REQUESTED** [review]
Great work! This is a nice improvement.

**Comment by jdm** [comment]
This should be ready to merge with those final fixes.

**Comment by jdm** [comment]
Go ahead and [squash](https://www.devroom.io/2011/07/05/git-squash-your-latests-commits-into-one/) all of the commits together, too.

**Pull request merged**